api4/plugin.go

```diff
@@ -338,39 +365,60 @@ func parseMarketplacePluginFilter(u *url.URL) (*model.MarketplacePluginFilter, e
 	}, nil
 }
 
-func installFromUrl(c *Context, w http.ResponseWriter, downloadUrl string, force bool) {
+func downloadFromUrl(c *Context, downloadUrl string) ([]byte, *model.AppError) {
 	if !model.IsValidHttpUrl(downloadUrl) {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
 	}
 
 	u, err := url.ParseRequestURI(downloadUrl)
 	if err != nil {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
 	}
 	if !*c.App.Config().PluginSettings.AllowInsecureDownloadUrl && u.Scheme != "https" {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.insecure_url.app_error", nil, "", http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.insecure_url.app_error", nil, "", http.StatusBadRequest)
 	}
 
 	client := c.App.HTTPService.MakeClient(true)
 	client.Timeout = INSTALL_PLUGIN_FROM_URL_HTTP_REQUEST_TIMEOUT
 
 	resp, err := client.Get(downloadUrl)
 	if err != nil {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.download_failed.app_error", nil, err.Error(), http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.download_failed.app_error", nil, err.Error(), http.StatusBadRequest)
 	}
 	defer resp.Body.Close()
 
 	fileBytes, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.reading_stream_failed.app_error", nil, err.Error(), http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.reading_stream_failed.app_error", nil, err.Error(), http.StatusBadRequest)
 	}
 
-	manifest, appErr := c.App.InstallPlugin(bytes.NewReader(fileBytes), force)
+	return fileBytes, nil
+}
+
+func verifyPlugin(c *Context, plugin *model.BaseMarketplacePlugin, pluginFileBytes []byte) ([]byte, *model.AppError) {
+	verified := false
+	verifiedSignature := []byte{}
+	for _, signature := range plugin.Signatures {
+		signatureBytes, err := base64.StdEncoding.DecodeString(signature.Signature)
+		if err != nil {
+			mlog.Error("Can't decode signature", mlog.String("public key hash", signature.PublicKeyHash), mlog.Err(err))
+		}
+		appErr := c.App.VerifyPlugin(bytes.NewReader(pluginFileBytes), bytes.NewReader(signatureBytes))
+		if appErr == nil {
+			verified = true
+			verifiedSignature = signatureBytes
```

Comment:
Should we just return at this point, and simplify the logic outside the loop to return an error?